src/demand/water_demand.jl

```diff
@@ -0,0 +1,934 @@
+"Update returnflow fraction for a non-irrigation water demand model"
+function return_flow_fraction!(model::NonIrrigationDemand)
+    (; returnflow_fraction) = model.variables
+    (; demand_gross, demand_net) = model.demand
+    @. returnflow_fraction = return_flow_fraction(demand_gross, demand_net)
+    return nothing
+end
+
+# return zero (gross water demand) if non-irrigation water demand sector is not defined
+return_flow_fraction!(model::NoNonIrrigationDemand) = nothing
+
+"Update water allocation for river and land domains based on local surface water (river) availability."
+function surface_water_allocation_local!(land_allocation, demand, river, network, dt)
+    (; surfacewater_alloc) = land_allocation.variables
+    (; surfacewater_demand) = demand.variables
+    (; act_surfacewater_abst_vol, act_surfacewater_abst, available_surfacewater) =
+        river.allocation.variables
+    # maps from the land domain to the internal river domain (linear index), excluding water bodies
+    index_river = network.land.index_river_wb
+    for i in eachindex(surfacewater_demand)
+        if index_river[i] > 0.0
+            # the available volume is limited by a fixed scaling factor of 0.8 to prevent
+            # rivers completely drying out. check for abstraction through inflow (external
+            # negative inflow) and adjust available volume.
+            if river.inflow[index_river[i]] < 0.0
+                inflow = river.inflow[index_river[i]] * dt
+                available_volume = max(river.volume[index_river[i]] * 0.80 + inflow, 0.0)
+            else
+                available_volume = river.volume[index_river[i]] * 0.80
+            end
+            # satisfy surface water demand with available local river volume
+            surfacewater_demand_vol = surfacewater_demand[i] * 0.001 * network.land.area[i]
+            abstraction_vol = min(surfacewater_demand_vol, available_volume)
+            act_surfacewater_abst_vol[index_river[i]] = abstraction_vol
+            # remaining available surface water and demand 
+            available_surfacewater[index_river[i]] =
+                max(available_volume - abstraction_vol, 0.0)
+            abstraction = (abstraction_vol / network.land.area[i]) * 1000.0
+            surfacewater_demand[i] = max(surfacewater_demand[i] - abstraction, 0.0)
+            # update actual abstraction from river and surface water allocation (land cell)
+            act_surfacewater_abst[index_river[i]] = abstraction
+            surfacewater_alloc[i] = abstraction
+        end
+    end
+    return nothing
+end
+
+"Update water allocation for river and land domains based on surface water (river) availability for allocation areas."
+function surface_water_allocation_area!(land_allocation, demand, river, network)
+    inds_river = network.river.indices_allocation_areas
+    inds_land = network.land.indices_allocation_areas
+    res_index = network.river.reservoir_index
+    lake_index = network.river.lake_index
```

Comment:
```suggestion
    (; indices_allocation_area, resevoir_index, lake_index) = network.river
    inds_land = network.land.indices_allocation_areas
```

The only disadvantage of this is that you have to use the exact name of the field

Reply:
Yes, and although a small disadvantage: the naming is quite different now between `inds_land` and  `indices_allocation_area`. Suggest to keep the original naming.